### PR TITLE
NMS-8142: Fix collapsible criteria is not persisted

### PR DESCRIPTION
**Pull request by pdzaccone**
* JIRA: https://issues.opennms.org/browse/NMS-8142

### CODE REVIEW

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java]: Can we not use `continue` statements and invert the if-statement and include the for block below, to make it more clear what is going on.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java]: Please always use brackets (`{´ and `}`) for each if.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java]: Please add the brackets at the end of the class/interface definition line.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: Please use a more meaningful name, like `criteria` or `searchCriteria`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmProviderDao.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmProviderDao.java]: We should name it `AlarmDaoProvider`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java]: Please remove TODO

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java]: the `createCriteria` method should set the id of the criteria based on the input's id. Then the next line can be removed.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java]: Bracket location should be at the end of the line before.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java]: Can the `createCriteria` method use `SearchResult` instead of `searchResult.getLabel()` instead?
The method could then also set the id.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java]: same as above.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProvider.java]: This should be named `CategoryProvider`

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java]: This should be named `CategoryDaoProvider`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java]: We should not have a default constructor.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProvider.java]: This should be called `IpInterfaceProvider`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProviderDao.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProviderDao.java]: This should be called `IpInterfaceDaoProvider`

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java]: Same as above. `createCriteria(String)` vs `createCriteria(SearchResult)`

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java]: see above.

**Review by mvrueden — CHANGES_REQUESTED** [review]

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java]: Can we add a logging statement if the `searchCriteria` was null?

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/test/java/org/opennms/features/topology/api/support/SavedHistoryTest.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/test/java/org/opennms/features/topology/api/support/SavedHistoryTest.java]: Can we add a comment to describe what the test is verifiing?

File: features/topology-map/org.opennms.features.topology.app/src/test/java/org/opennms/features/topology/app/internal/IpLikeSearchProviderTest.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/test/java/org/opennms/features/topology/app/internal/IpLikeSearchProviderTest.java]: Bracket should be at the end of the line before (-:

File: features/topology-map/plugins/org.opennms.features.topology.plugins.ncs/src/main/java/org/opennms/features/topology/plugins/ncs/NCSSearchProvider.java
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.ncs/src/main/java/org/opennms/features/topology/plugins/ncs/NCSSearchProvider.java]: Should only implement HistoryAwareSearchProvider.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: This looks fine to me, but why do we have a `CategoryProviderTest` in addition to this? We should remove the test/dummy implementation. This is true for all test provideres if they are not used anywhere else.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: I think we should actually return `capturedCriteria.getValues()` here.

**Review by mvrueden — CHANGES_REQUESTED** [review]
This looks very good now 👍 
I found some more minor issues. 
Please address them. 
I will give this a spin and verify that it actually still works and also that the history looks good

**Comment by mvrueden** [comment]
I created a Pull Request to make the "login" and "logout" methods available to all tests. See https://github.com/OpenNMS/smoke-tests/pull/2 . We have to wait until it gets merged, though.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java
Comment (pdzaccone) [reply features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: `capturedCriteria` object was introduced to intercept parameters from the method `addCriteria`. Objects from the `capturedCriteria` are later used in assertion.
If the method `getCriteria` returns the same values, then the comparison will always return true. The very idea of test will then be lost.
Besides, at the beginning the `capturedCriteria` is empty and the `getCriteria` method is called earlier than `addCriteria`

Comment (mvrueden) [reply features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: Then let me rephrase. In order that the test actually makes sense, we should mock the `addCriteria()`, `removeCriteria()` and `getCriteria()` methods, rather than always return a fixed list. Otherwise the whole purpose of the test is questionable.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java:206]: Can we add a description here on what we are doing and why?

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java]: Can we add a description here to help developers understand what the `SearchCriteria` is supposed to do?

File: smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java]: This can be removed now

File: smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java]: This can be removed now

**Review by mvrueden — APPROVED** [review]

**Comment by mvrueden** [comment]
Should be merged to opennms/foundation-2017. Closing.

**Pull request closed without merging**
